prime-dotnet-webapi/Services/DefaultAutomaticAdjudicationService.cs

```diff
@@ -0,0 +1,213 @@
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Prime.Models;
+
+namespace Prime.Services
+{
+    public class DefaultAutomaticAdjudicationServiceService : BaseService, IAutomaticAdjudicationService
+    {
+        readonly List<IAutomaticAdjudicationRule> _rules = new List<IAutomaticAdjudicationRule>();
+
+        public DefaultAutomaticAdjudicationServiceService(
+            ApiDbContext context, IHttpContextAccessor httpContext)
+            : base(context, httpContext)
+        {
+            // add the rules to the list that this implementation will process
+            _rules.Add(new SelfDeclarationRule());
+            _rules.Add(new AddressRule());
+            _rules.Add(new PumpProviderRule());
+            _rules.Add(new CertificationNameRule());
+            _rules.Add(new LicenceClassRule());
+            _rules.Add(new LicenceNumberRule());
+        }
+
+        public bool QualifiesForAutomaticAdjudication(Enrolment enrolment)
+        {
+            // Check all of the rules to see if this should qualify to be automatically adjudicated.
+            // If it does not qualify not, it should add the reasons it did not meet the criteria to the current enrolment status
+            bool passed = true;
+            foreach (var rule in _rules)
+            {
+                passed &= rule.ProcessRule(enrolment);
+            }
+
+            return passed;
+        }
+
+        private interface IAutomaticAdjudicationRule
+        {
+            bool ProcessRule(Enrolment enrolment);
+        }
+
+        private abstract class BaseAutomaticAdjudicationRule : IAutomaticAdjudicationRule
+        {
+            public bool ProcessRule(Enrolment enrolment)
+            {
+                // make sure that there is an enrolment to process the rule against
+                if (enrolment == null)
+                {
+                    throw new ArgumentNullException(nameof(enrolment), "Could not process enrolment rule, passed in Enrolment cannot be null.");
+                }
+
+                // process the rule and check the results
+                var results = this.ProcessRuleInternal(enrolment);
+
+                bool ruleFailed = results.Any();
+
+                if (ruleFailed)
+                {
+                    // get the current status record for the enrolment
+                    var currentStatus = enrolment.CurrentStatus;
+
+                    // make sure there is a current status
+                    if (currentStatus == null)
+                    {
+                        throw new InvalidOperationException($"Could not process enrolment rule, current status was missing for Enrolment.Id={enrolment.Id}.");
+                    }
+
+                    // for every item returned in the results, add the reason to the current status
+                    foreach (var item in results)
+                    {
+                        if (currentStatus.EnrolmentStatusReasons == null)
```

Comment:
Looks like we could pull this out of the foreach loop, just for clarity.

Reply:
I pulled it out of the loop - I had this both ways before as I didn't want to create it is there were no reasons, but I don't think it will really hurt to make an empty list.